backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/ExtractionRequirements.java

```diff
@@ -184,4 +195,54 @@ else if ( !requiredStoredFields.isEmpty() ) {
 			return new StoredFieldsValuesDelegate.Factory( storedFieldVisitor, requiredNestedDocumentPathsForStoredFields );
 		}
 	}
+
+	private static final class HibernateSearchTopScoreDocCollectorManager extends TopScoreDocCollectorManager {
+		private final int numHits;
+		private final int offset;
+		private final Integer limit;
+
+		public HibernateSearchTopScoreDocCollectorManager(int offset, Integer limit, int numHits, ScoreDoc after,
+				int totalHitsThreshold, boolean supportsConcurrency) {
+			super( numHits, after, totalHitsThreshold, supportsConcurrency );
+			this.numHits = numHits;
+			this.offset = offset;
+			this.limit = limit;
+		}
+
+		@Override
+		public TopDocs reduce(Collection<TopScoreDocCollector> collectors) {
+			final TopDocs[] topDocs = new TopDocs[collectors.size()];
+			int i = 0;
+			for ( TopScoreDocCollector collector : collectors ) {
+				topDocs[i++] = collector.topDocs();
+			}
+			return TopDocs.merge( offset, limit == null ? numHits : Math.min( numHits, offset + limit ), topDocs );
+		}
+	}
+
+	private static final class HibernateSearchTopFieldCollectorManager extends TopFieldCollectorManager {
+		private final int numHits;
+		private final Sort sort;
+		private final int offset;
+		private final Integer limit;
+
+		public HibernateSearchTopFieldCollectorManager(int offset, Integer limit, Sort sort, int numHits, FieldDoc after,
+				int totalHitsThreshold, boolean supportsConcurrency) {
+			super( sort, numHits, after, totalHitsThreshold, supportsConcurrency );
+			this.numHits = numHits;
+			this.offset = offset;
+			this.limit = limit;
+			this.sort = sort;
+		}
+
+		@Override
+		public TopFieldDocs reduce(Collection<TopFieldCollector> collectors) {
+			final TopFieldDocs[] topDocs = new TopFieldDocs[collectors.size()];
+			int i = 0;
+			for ( TopFieldCollector collector : collectors ) {
+				topDocs[i++] = collector.topDocs();
+			}
+			return TopDocs.merge( sort, offset, limit == null ? numHits : Math.min( numHits, offset + limit ), topDocs );
```

Comment:
As far as I can see, `numHits` here is equal to `maxDocs` in `LuceneSearcherImpl`, and that is already defined as:

```
		int maxDocs = getMaxDocs( indexSearcher.getIndexReader(), offset, limit );
```

Which is:

```
	private int getMaxDocs(IndexReader reader, int offset, Integer limit) {
		if ( limit == null ) {
			return reader.maxDoc();
		}
		else if ( limit.equals( 0 ) ) {
			return 0;
		}
		else {
			return Math.min( offset + limit, reader.maxDoc() );
		}
	}
```

So... I don't think you need to carry around the limit all the way down. Just using `numHits` here should lead to the same results... I think?